march_hardware/src/IMotionCube.cpp

```diff
@@ -331,6 +344,33 @@ void IMotionCube::reset()
   sdo_bit16_write(this->slaveIndex, 0x2080, 0, 1);
 }
 
+uint32_t IMotionCube::computeSWCheckSum(int& start_address, int& end_address)
+{
+  size_t pos = 0;
+  size_t old_pos = 0;
+  uint16_t sum = 0;
+  std::string delimiter = "\n";
+  std::string token;
+  while ((pos = sw_stream_.find(delimiter, old_pos)) != std::string::npos)
+  {
+    token = sw_stream_.substr(old_pos, pos - old_pos);
+    if (old_pos == 0)
+    {
+      start_address = std::stoi(token, nullptr, 16);
+      token = "0";
+    }
+    if (pos - old_pos < 2)  // delimiter has length of 1 two \n in a row has difference in positions of 1
+    {
+      end_address = end_address + start_address - 2;  // The -2 compensates the offset of the end_address
+      return sum;
+    }
+    end_address++;
+    sum += std::stoi(token, nullptr, 16);  // State that we are looking at hexadecimal numbers
+    old_pos = pos + 1;                     // Make sure to check the position after the previous one
+  }
+  return sum;  // Only reached if something goes wrong and the code doesn't enter the while loop
```

Comment:
Perhaps you should throw an exception here instead, otherwise it would be impossible to know for the caller that the method failed. You can use the `HardwareException` and create a custom `ErrorType` with it. See `error/error_types.h` and `error/error_types.cpp` for examples.